stats/postgres_data_access.go

```diff
@@ -0,0 +1,186 @@
+}
+
+// AddPatientStat increments the relevant county and subdivision statistics
+// based on the given patient's address.
+func (da *PgStatsDataAccess) AddPatientStat(patient *models.Patient) (err error) {
+
+	countyfp, cousubfp, err := da.identifyCountyAndSubdivisionForPatient(patient)
+
+	if err != nil {
+		return
+	}
+
+	if patientGenderIsValid(patient) {
+		err = da.updateStats(countyfp, cousubfp, patient.Gender, "increment")
+
+	} else {
+		err = InvalidGenderError
+	}
+	return
+}
+
+// RemovePatientStat decrements the relevant county and subdivision statistics
+// based on the given patient's address.
+func (da *PgStatsDataAccess) RemovePatientStat(patient *models.Patient) (err error) {
+
+	countyfp, cousubfp, err := da.identifyCountyAndSubdivisionForPatient(patient)
+
+	if err != nil {
+		return
+	}
+
+	if patientGenderIsValid(patient) {
+		err = da.updateStats(countyfp, cousubfp, patient.Gender, "decrement")
+
+	} else {
+		err = InvalidGenderError
+	}
+	return
+}
+
+// AddConditionStat increments the relevant county and subdivision condition facts
+// based on the given patient's address.
+func (da *PgStatsDataAccess) AddConditionStat(patient *models.Patient, condition *models.Condition) (err error) {
+
+	countyfp, cousubfp, err := da.identifyCountyAndSubdivisionForPatient(patient)
+	diseasefp, err := da.identifyDiseaseForCondition(condition)
+
+	if err != nil {
+		return
+	}
+
+	if patientGenderIsValid(patient) {
+		err = da.updateFacts(countyfp, cousubfp, diseasefp, patient.Gender, "increment")
+
+	} else {
+		err = InvalidGenderError
+	}
+	return
+}
+
+// RemoveConditionStat decrements the relevant county and subdivision condition facts
+// based on the given patient's address.
+func (da *PgStatsDataAccess) RemoveConditionStat(patient *models.Patient, condition *models.Condition) (err error) {
+
+	countyfp, cousubfp, err := da.identifyCountyAndSubdivisionForPatient(patient)
+	diseasefp, err := da.identifyDiseaseForCondition(condition)
+
+	if err != nil {
+		return
+	}
+
+	if patientGenderIsValid(patient) {
+		err = da.updateFacts(countyfp, cousubfp, diseasefp, patient.Gender, "decrement")
+
+	} else {
+		err = InvalidGenderError
+	}
+	return
+}
+
+// identifyCountyAndSubdivisionForPatient returns the countyfp and cousubfp that
+// match the subdivision in the given patient's address.
+func (da *PgStatsDataAccess) identifyCountyAndSubdivisionForPatient(patient *models.Patient) (countyfp, cousubfp string, err error) {
+	err = da.DB.QueryRow("SELECT countyfp, cousubfp FROM tiger.cousub WHERE name = $1", patient.Address[0].City).Scan(&countyfp, &cousubfp)
+	return
+}
+
+// identifyDiseaseForCondition returns the diseasefp that matches the given
+// condition's SNOMED code, if any.
+func (da *PgStatsDataAccess) identifyDiseaseForCondition(condition *models.Condition) (diseasefp string, err error) {
+	err = da.DB.QueryRow("SELECT diseasefp FROM synth_ma.synth_disease WHERE code_snomed = $1", getSnomedCode(condition)).Scan(&diseasefp)
```

Comment:
In terms of multiple codes for one stat, I did this:

Each row in the fact table has two columns, "stat_name" and "condition_name". Condition name must be unique but stat name is not. So for example, if the conditions "pre_diabetes" and "diabetes" both fall under the "diabetes" statistic, then querying our fact tables for the "diabetes" stat would need to return the aggregate of multiple rows. But this allows us to register many different snomed codes and unique conditions to represent one "statistic" that we care about.